KAFKA-9039: Optimize ReplicaFetcher fetch path

lbradstreet opened the pull request:
Improves the performance of the replica fetcher for high partition count fetch requests, where a majority of the partitions did not update between fetch requests. All benchmarks were run on an r5x.large.

Vanilla
Benchmark                                  (partitionCount)  Mode  Cnt        Score       Error  Units
ReplicaFetcherThreadBenchmark.testFetcher               100  avgt   15    26491.825 ±   438.463  ns/op
ReplicaFetcherThreadBenchmark.testFetcher               500  avgt   15   153941.952 ±  4337.073  ns/op
ReplicaFetcherThreadBenchmark.testFetcher              1000  avgt   15   339868.602 ±  4201.462  ns/op
ReplicaFetcherThreadBenchmark.testFetcher              5000  avgt   15  2588878.448 ± 22172.482  ns/op

From 100 to 5000 partitions the latency increase is 2588878.448 / 26491.825 = 97.

Avoid gettimeofdaycalls in steady state fetch states
854588804d7cf96a10d6e4a74db533deb7e26b55

Benchmark                                  (partitionCount)  Mode  Cnt        Score      Error  Units
ReplicaFetcherThreadBenchmark.testFetcher               100  avgt   15    22685.381 ±  267.727  ns/op
ReplicaFetcherThreadBenchmark.testFetcher               500  avgt   15   113622.521 ± 1854.254  ns/op
ReplicaFetcherThreadBenchmark.testFetcher              1000  avgt   15   273698.740 ± 9269.554  ns/op
ReplicaFetcherThreadBenchmark.testFetcher              5000  avgt   15  2189223.207 ± 1706.945  ns/op

From 100 to 5000 partitions the latency increase is 2189223.207 / 22685.381 = 97X

Avoid copying partition states to maintain fetch offsets 
29fdd6094b828153c762f1d99b645481f7200cee

Benchmark                                  (partitionCount)  Mode  Cnt        Score      Error  Units
ReplicaFetcherThreadBenchmark.testFetcher               100  avgt   15    17039.989 ±  609.355  ns/op
ReplicaFetcherThreadBenchmark.testFetcher               500  avgt   15    99371.086 ± 1833.256  ns/op
ReplicaFetcherThreadBenchmark.testFetcher              1000  avgt   15   216071.333 ± 3714.147  ns/op
ReplicaFetcherThreadBenchmark.testFetcher              5000  avgt   15  2035678.223 ± 5195.232  ns/op

From 100 to 5000 partitions the latency increase is 2035678.223 / 17039.989 = 119X

Keep lag alongside PartitionFetchState to avoid expensive isReplicaInSync check
0e57e3e725ff70ad8845681b02873229c3c64544

Benchmark                                  (partitionCount)  Mode  Cnt        Score      Error  Units
ReplicaFetcherThreadBenchmark.testFetcher               100  avgt   15    15131.684 ±  382.088  ns/op
ReplicaFetcherThreadBenchmark.testFetcher               500  avgt   15    86813.843 ± 3346.385  ns/op
ReplicaFetcherThreadBenchmark.testFetcher              1000  avgt   15   193050.381 ± 3281.833  ns/op
ReplicaFetcherThreadBenchmark.testFetcher              5000  avgt   15  1801488.513 ± 2756.355  ns/op

From 100 to 5000 partitions the latency increase is 1801488.513 / 15131.684 = 119X

Fetch session optimizations (mostly presizing the next hashmap, and avoiding making a copy of sessionPartitions, as a deep copy is not required for the ReplicaFetcher)
2614b2417a6aac87e5decbae0a29d4f37ba9c7e3

Benchmark                                  (partitionCount)  Mode  Cnt        Score      Error  Units
ReplicaFetcherThreadBenchmark.testFetcher               100  avgt   15    11386.203 ±  416.701  ns/op
ReplicaFetcherThreadBenchmark.testFetcher               500  avgt   15    60820.292 ± 3163.001  ns/op
ReplicaFetcherThreadBenchmark.testFetcher              1000  avgt   15   146242.158 ± 1937.254  ns/op
ReplicaFetcherThreadBenchmark.testFetcher              5000  avgt   15  1366768.926 ± 3305.712  ns/op

From 100 to 5000 partitions the latency increase is 1366768.926 / 11386.203 = 120

lbradstreet committed:
Add replica fetcher benchmark

lbradstreet committed:
Avoid gettimeofdaycalls in steady state fetch states

lbradstreet committed:
Avoid copying partition states to maintain fetch offsets
We already have the partition fetch data in the fetch session, so a copy
is not required.

lbradstreet committed:
Keep lag alongside PartitionFetchState to avoid expensive
isReplicaInSync check. Reduce cost of updating map by not unnecessarily
wrapping in ClientIdTopicPartition in FetcherStats

lbradstreet committed:
Fix logging of currentLeaderEpoch

lbradstreet committed:
Fetch session optimizations

lbradstreet commented:
According to profiling, a majority of the rest of the time goes to looking up the logStartOffset for each partition on each pass (https://github.com/apache/kafka/pull/7443/files#diff-a8437f241a5ae585d5805ee769313080R252), and updating the incremental fetch session (https://github.com/apache/kafka/pull/7443/files#diff-a8437f241a5ae585d5805ee769313080R253). If we can drop those further, we may be able to save another 50-75%.

lbradstreet commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -702,29 +705,21 @@ class FetcherLagMetrics(metricId: ClientIdTopicPartition) extends KafkaMetricsGr
-      fetcherLagMetrics.lag <= 0
-    else
-      false
+    stats.getAndMaybePut(topicPartition)
```
Wrapping each put in a ClientIdTopicPartition served no purpose here, as metricId is fixed for every instantiation of FetcherLagStats. This adds additional object creation/GC overhead, as well as additional hashing overhead.

lbradstreet commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -307,8 +309,8 @@ abstract class AbstractFetcherThread(name: String,
-            val fetchState = fetchStates(topicPartition)
-            if (fetchState.fetchOffset == currentFetchState.fetchOffset && currentFetchState.isReadyForFetch) {
+            val fetchPartitionData = sessionPartitions.get(topicPartition)
+            if (fetchPartitionData != null && fetchPartitionData.fetchOffset == currentFetchState.fetchOffset && currentFetchState.isReadyForFetch) {
```
This could be cleaned up a little.

lbradstreet commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -236,15 +241,15 @@ class ReplicaFetcherThread(name: String,
     val partitionsWithError = mutable.Set[TopicPartition]()
 
-    val builder = fetchSessionHandler.newBuilder()
+    val builder = fetchSessionHandler.newBuilder(partitionMap.size, false)
```
Use the new builder which presizes the PartitionData hashmap, and which does not make another copy of sessionPartitions. We will not be modifying sessionPartitions, or using it again after we have processed the fetch request, so we do not need to make an unnecessary copy.

lbradstreet commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -265,7 +270,7 @@ class ReplicaFetcherThread(name: String,
         .toForget(fetchData.toForget)
         .metadata(fetchData.metadata)
-      Some(requestBuilder)
+      Some(ReplicaFetch(fetchData.sessionPartitions(), requestBuilder))
```
By passing through sessionPartitions we can avoid making another copy of the PartitionMap in https://github.com/apache/kafka/pull/7443/files#diff-2d03a5d4349a23e56c09b4e773055249L119. sessionPartitions contains all of the PartitionData that we need to process the fetch response.

lbradstreet commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -330,9 +335,8 @@ class ReplicaFetcherThread(name: String,
-    val isReplicaInSync = fetcherLagStats.isReplicaInSync(topicPartition)
-    !isReplicaInSync && quota.isThrottled(topicPartition) && quota.isQuotaExceeded
+  private def shouldFollowerThrottle(quota: ReplicaQuota, fetchState: PartitionFetchState, topicPartition: TopicPartition): Boolean = {
+    !fetchState.isReplicaInSync && quota.isThrottled(topicPartition) && quota.isQuotaExceeded
```
We avoid a HashMap lookup here by including the lag in the PartitionFetchState directly.

lbradstreet commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -768,21 +763,24 @@ object PartitionFetchState {
   def isTruncating: Boolean = state == Truncating && !isDelayed
 
-  def isDelayed: Boolean = delay.getDelay(TimeUnit.MILLISECONDS) > 0
+  def isDelayed: Boolean = delay.exists(_.getDelay(TimeUnit.MILLISECONDS) > 0)
```
Avoid an unnecessary gettimeofday when partitions are in Fetching state (the common case) by using an Option[DelayedItem].

guozhangwang commented on jmh-benchmarks/src/main/java/org/apache/kafka/jmh/fetcher/ReplicaFetcherThreadBenchmark.java:
```diff
@@ -0,0 +1,314 @@
+        // so that we do not measure this time as part of the steady state work
+        fetcher.doWork();
+        // handle response to engage the incremental fetch session handler
+        fetcher.fetchSessionHandler().handleResponse(new FetchResponse<>(Errors.NONE, initialFetched, 0, 999));
```
Do we need to call handleResponse? I thought it's triggered in `doWork()` already.

lbradstreet replied:
We've overridden fetchFromLeader so this path is not exercised (https://github.com/apache/kafka/blob/d2a13d82702e0e0f0b2275b984b4babdbd976ae3/jmh-benchmarks/src/main/java/org/apache/kafka/jmh/fetcher/ReplicaFetcherThreadBenchmark.java#L310)

guozhangwang replied:
Makes sense.

guozhangwang commented:
Thanks for the combo-patch @lbradstreet ! I think this is very helpful. A couple of thoughts on top of my head:

1. in the baseline, from 100 to 5000 partitions the latency increased by about 97X; in the PR that contains all optimizations, although the absolute latency is reduced, from 100 to 5000 the relative ratio is actually 120X, which is a bit weird. Ideally I'd expect we can achieve close to 50X, i.e. linear to num.partitions.

2. We should probably look into all the INFO level log4j entries to avoid any expensive evaluations, e.g. the one below

```
info(s"Beginning/resuming copy of partition $tp from offset ${fetchState.fetchOffset}. " +
        s"Including this partition, there are ${partitionMap.size} remaining partitions to copy by this thread.")
```

partitionMap.size in Scala is O(n), although the passed-in map is a LinkedHashMap whose size call is O(1). Honestly I do not know which one would be triggered, but worth reducing it to be safe.

3. One bold idea I have is that in the `SessionHandler`, we can be even smarter, e.g. to not include all the partitions if we found that the previous request only returns data for N partitions out of all, then in the next buildFetch we could just fetch for the next N partitions.

guozhangwang commented:
For 1) I've just updated the description to include the relative ratio of each optimization, it seems that 29fdd60 makes it less scalable with the num.partitions since it's only useful with small num.partitions.

guozhangwang commented:
4. Another wild thought: in `doWork` we can comment out `maybeTruncate` and see if that alone is taking up long time compared with the baseline, since in practice `maybeTruncate` should mostly be no-op, and if itself is taking lots of time we can also make another thorough pass on that function. I can do the benchmark taking from your branch if you think it's a good idea.

lbradstreet commented:
> For 1) I've just updated the description to include the relative ratio of each optimization, it seems that [29fdd60](https://github.com/apache/kafka/commit/29fdd6094b828153c762f1d99b645481f7200cee) makes it less scalable with the num.partitions since it's only useful with small num.partitions.

Thanks! I'm assuming that you're OK with it still, because it still helps across the board? It just helps less for the high partition count case?

lbradstreet committed:
Avoid unecessary sessionPartitions check. New strategy ensures that only
new partitions are left in next map.

lbradstreet commented:
> Is it always better to construct a new map `updated` than updating `next` directly?

It's a good question, and I'm not sure. The idea was that by making next smaller we could make the lookups progressively quicker as we process sessionsPartitions. It did seem to improve performance for our benchmark. I forgot to include the second optimization that it allows us to make https://github.com/apache/kafka/pull/7443/commits/f0bf23acd8198eae0faed67c2414301eccf04dc6. This helps reduce containsKey checks, and is worth keeping if we leave the change to keep the second map in. If not we will need to remove it again.

guozhangwang commented:
> Thanks! I'm assuming that you're OK with it still, because it still helps across the board? It just helps less for the high partition count case?

Yeah that's okay with me still. Just trying to point it out that with large number of partitions there seems other dominating factors.

lbradstreet committed:
Minimize differences in fetch session handler vs trunk
We will get better performance improvements by redesigning how
incremental fetch sessions work in the Replica Fetcher. Minimizing
differences in the fetch session will minimize risks of the changeset.

lbradstreet commented:
@guozhangwang I removed (https://github.com/apache/kafka/pull/7443/commits/32e30ace5865e0a5b769cad626f64bcdf5091faf) most of the risky FetchSessionHandler changes that only showed up in the FetchSessionHandlerBenchmark, as I think we should redesign how the Replica Fetcher incremental fetch session is maintained, rather than try and optimize it further. 

I re-ran the benchmark on the r5x.large and it shows very little differences vs the previous commit.

```
Benchmark                                  (partitionCount)  Mode  Cnt        Score      Error  Units
ReplicaFetcherThreadBenchmark.testFetcher               100  avgt   15    11534.716 ±  327.903  ns/op
ReplicaFetcherThreadBenchmark.testFetcher               500  avgt   15    58781.785 ± 2209.587  ns/op
ReplicaFetcherThreadBenchmark.testFetcher              1000  avgt   15   151653.533 ±  400.809  ns/op
ReplicaFetcherThreadBenchmark.testFetcher              5000  avgt   15  1359152.056 ±  912.209  ns/op
```

lbradstreet committed:
Spotbugs fixes

lbradstreet commented on jmh-benchmarks/src/main/java/org/apache/kafka/jmh/fetcher/ReplicaFetcherThreadBenchmark.java:
```diff
@@ -0,0 +1,313 @@
+        }
+
+        @Override
+        public Seq<Tuple2<TopicPartition, FetchResponse.PartitionData<Records>>> fetchFromLeader(FetchRequest.Builder fetchRequest) {
```
Unfortunately this line breaks for scala 2.13 due to https://docs.scala-lang.org/overviews/core/collections-migration-213.html#option-1-migrate-back-to-scalacollectionseq. I'm not how to fix this in Java land without breaking the 2.11 and 2.12 builds. The compat libraries only seem to be usable from scala.

guozhangwang replied:
Actually do we need the return type to be `Seq`, than just a `Map`? It seems we do not need to rely on any keyset ordering at all in the remaining handling logic.

lbradstreet replied:
Changing the return type for fetchFromLeader is my preferred fix. There's no particular reason to convert it to a seq, other than to match the ReplicaAlterLogDirsThread implementation, which could easily be converted to return a map instead (https://github.com/apache/kafka/blob/0e57e3e725ff70ad8845681b02873229c3c64544/core/src/main/scala/kafka/server/ReplicaAlterLogDirsThread.scala#L80).

guozhangwang replied:
SGTM!

lbradstreet replied:
Fixed by https://github.com/apache/kafka/pull/7443/commits/1a0ed01d9d90b4b3310cd76944f644c840e70217

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -185,7 +185,18 @@ public String toString() {
          * incremental fetch requests (see below).
          */
-        private LinkedHashMap<TopicPartition, PartitionData> next = new LinkedHashMap<>();
+        private LinkedHashMap<TopicPartition, PartitionData> next;
```
I'd suggest we make one step further, to just make the `sessionPartitions` a local variable only: this is because in later handling logic after the response is sent back, we only need `sessionPartitions.keySet / size` only, which we could maintain separately as a member field which is populated after the build() call. WDYT?

guozhangwang replied:
NVM I realized now you also need the `sessionPartitions` for saving one fetch state copying as well.

guozhangwang reviewed (commented):
LGTM. I do not have further comments except the fix for scala 2.13 above. I will try to run the benchmark included in your PR on my local MP5 as well to confirm its effects before merging.

guozhangwang commented:
I've run the benchmark from this PR on my laptop and here's the results:

Baseline (trunk)

```
ReplicaFetcherThreadBenchmark.testFetcher               100  avgt   15    19845.007 ±    95.308  ns/op
ReplicaFetcherThreadBenchmark.testFetcher               500  avgt   15   131247.995 ±   777.615  ns/op
ReplicaFetcherThreadBenchmark.testFetcher              1000  avgt   15   290156.189 ±  1590.831  ns/op
ReplicaFetcherThreadBenchmark.testFetcher              5000  avgt   15  2401985.724 ± 45898.838  ns/op
```

For 100/500/1000/5000 partitions, the cost scale factor (i.e. score per partition) increased from 1 to 1/1.32/1.46/2.42

This PR:

```
ReplicaFetcherThreadBenchmark.testFetcher               100  avgt   15     9480.761 ±   303.979  ns/op
ReplicaFetcherThreadBenchmark.testFetcher               500  avgt   15    67359.693 ±  4235.383  ns/op
ReplicaFetcherThreadBenchmark.testFetcher              1000  avgt   15   155223.743 ±  2936.436  ns/op
ReplicaFetcherThreadBenchmark.testFetcher              5000  avgt   15  1297303.182 ± 11500.040  ns/op
```

The latency reduction ranges between 65% to 50%.

Interesting, for 100/500/1000/5000 partitions, the cost scale factor (i.e. score per partition) increased from 1 to 1/1.42/1.64/2.72, which is higher than the baseline. I means the optimizations we've done in this PR benefits the scenario with less partitions more than larger num.partitions.

Overall, I still think we can merge the PR as-is and continue the next optimizations on top of it.

junrao commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -768,21 +763,24 @@ object PartitionFetchState {
       s", currentLeaderEpoch=$currentLeaderEpoch" +
       s", state=$state" +
-      s", delay=${delay.delayMs}ms" +
+      s", delay=${delay.map(_.delayMs).getOrElse(0)}ms" +
```
Could we add the newly added field lag?

lbradstreet replied:
Good suggestion, thanks. Resolved by https://github.com/apache/kafka/pull/7443/commits/1aa07f40ac1c69349f51604ba85e44344e1644d2

lbradstreet committed:
Fix compile error on scala 2.13 by making fetchFromLeader return a map

lbradstreet committed:
Pre-compute or pass through lags through fetch states where possible

lbradstreet commented:
> Is it always better to construct a new map `updated` than updating `next` directly?

Resolved by https://github.com/apache/kafka/pull/7443/commits/32e30ace5865e0a5b769cad626f64bcdf5091faf

lbradstreet commented:
Build failures are due to broken trunk (https://github.com/apache/kafka/pull/7521)

lbradstreet commented:
retest this please

guozhangwang commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -281,7 +281,7 @@ abstract class AbstractFetcherThread(name: String,
                                   fetchRequest: FetchRequest.Builder): Unit = {
     val partitionsWithError = mutable.Set[TopicPartition]()
-    var responseData: Seq[(TopicPartition, FetchData)] = Seq.empty
+    var responseData: Map[TopicPartition, FetchData] = Map.empty
```
+100

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -236,15 +241,15 @@ class ReplicaFetcherThread(name: String,
     val partitionsWithError = mutable.Set[TopicPartition]()
 
-    val builder = fetchSessionHandler.newBuilder()
+    val builder = fetchSessionHandler.newBuilder(partitionMap.size, false)
```
Hmm, here, we choose not to copy sessionPartitions in the builder. However, in line 265 and line 273, we still read  sessionPartitions. Is that correct?

lbradstreet replied:
I believe this is still correct. We do not copy sessionPartitions, but sessionPartitions will also not be modified until the next fetch/doWork iteration, and all use of it is from the same thread. I believe it is thus safe to use it without making a copy.

junrao replied:
Thanks for the explanation. That makes sense.

junrao commented on core/src/main/scala/kafka/server/ReplicaAlterLogDirsThread.scala:
```diff
@@ -34,7 +35,7 @@ import org.apache.kafka.common.requests.FetchResponse.PartitionData
 
 import scala.collection.JavaConverters._
-import scala.collection.{Map, Seq, Set, mutable}
+import scala.collection.{mutable, Map, Seq, Set}
```
It seems that Seq is unused now.

lbradstreet replied:
I believe it's still used here https://github.com/apache/kafka/pull/7443/files/4a306ad7967e20712ec2c0ea085a91b5f471b9a3#diff-6446cfed30fa23cfc249b9537fffbe35R79. It's not showing up for me as a checkstyle failure.

junrao commented on core/src/test/scala/unit/kafka/server/AbstractFetcherThreadTest.scala:
```diff
@@ -25,6 +25,7 @@ import com.yammer.metrics.Metrics
 import kafka.cluster.BrokerEndPoint
 import kafka.log.LogAppendInfo
 import kafka.message.NoCompressionCodec
+import kafka.server.AbstractFetcherThread.ReplicaFetch
```
unused import

lbradstreet replied:
I believe it's used here https://github.com/apache/kafka/pull/7443/files/4a306ad7967e20712ec2c0ea085a91b5f471b9a3#diff-f84579dc9ecc952780289986cddecc46R905. It's not showing up as a checkstyle error for me.

junrao reviewed (commented):
@lbradstreet : Thanks for the PR. Looks good overall. A few comments below.

lbradstreet committed:
Output lag in PartitionFetchState toString

junrao reviewed (approved):
@lbradstreet : Thanks for the latest PR. LGTM

guozhangwang commented:
LGTM. I will merge after the jenkins job completes.

lbradstreet commented:
Failures in:
2.12
```
kafka.api.SaslSslAdminClientIntegrationTest.testCreateTopicsResponseMetadataAndConfig
kafka.log.LogCleanerIntegrationTest.testIsThreadFailed
org.apache.kafka.streams.integration.KTableKTableForeignKeyJoinIntegrationTest.doLeftJoinFilterOutRapidlyChangingForeignKeyValues
org.apache.kafka.streams.integration.KTableKTableForeignKeyJoinIntegrationTest.doInnerJoinFilterOutRapidlyChangingForeignKeyValues
```

2.13:
```
kafka.admin.ReassignPartitionsClusterTest.shouldTriggerReassignmentWithZnodePrecedenceOnControllerStartup
kafka.api.SaslSslAdminClientIntegrationTest.testCreateTopicsResponseMetadataAndConfig
```
